### PR TITLE
feat: A solution for SecretRef (BasicAuthRef )

### PR DESCRIPTION
**Pull request by yuwenma**
For sensitive data in direct resource API, we want to reference a k8s Secret. 

This PR is diff base from #3042.

### CODE REVIEW

File: pkg/controller/direct/bigqueryconnection/connection_mapping.go
Comment (justinsb) [inline pkg/controller/direct/bigqueryconnection/connection_mapping.go:298]: Does this cause the round-trip tests to fail?

File: pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/awsconnectionbasic/_http.log
Comment (justinsb) [inline pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/awsconnectionbasic/_http.log:97]: What happened to this field?

**Comment by justinsb** [comment]
Some suggestions but I don't think they are blocking (they don't impact external behaviour).

/approve
/lgtm

/hold for suggestions

File: pkg/controller/direct/bigqueryconnection/connection_controller.go
Comment (yuwenma) [reply pkg/controller/direct/bigqueryconnection/connection_controller.go:97]: +1. "Maybe it should be part of the interface though," Before it happens, let me change this to private. Thanks!

File: apis/refs/v1beta1/secret/basicauth.go
Comment (yuwenma) [reply apis/refs/v1beta1/secret/basicauth.go:33]: Agree to not export (that's also what I did at the beginning). But when this object is referenced and parsed in `runtime.DefaultUnstructuredConverter.FromUnstructured`, it panics. Here's the related k/k issue.  
Here's the related issue https://github.com/kubernetes/kubernetes/issues/124154

File: pkg/controller/direct/bigqueryconnection/connection_mapping.go
Comment (yuwenma) [reply pkg/controller/direct/bigqueryconnection/connection_mapping.go:298]: We don't have roundtrip test for this resource yet. I think we cannot not map input-only field back to KRM any ways, can we?

File: pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/awsconnectionbasic/_http.log
Comment (yuwenma) [reply pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/awsconnectionbasic/_http.log:97]: This PR based on a pending PR. #3042 (implements the update field mask in MockGCP)
I'll rebase once that other one merges. This should go away

File: apis/refs/v1beta1/secret/basicauth.go
Comment (justinsb) [reply apis/refs/v1beta1/secret/basicauth.go:33]: Ah - good to know.  Disappointing that we haven't fixed that!

Comment (justinsb) [reply apis/refs/v1beta1/secret/basicauth.go:33]: I guess we can image that we _might_ let users set the raw values directly, and then having these fields makes sense.

We don't want to let users set the raw values, because we want to nudge them to using a secret.  But if we get push-back, we could change our minds here and allow the values to be set via the json tag.  And then if we did that, this structure and normalization _would_ make sense.

File: pkg/controller/direct/bigqueryconnection/connection_mapping.go
Comment (justinsb) [reply pkg/controller/direct/bigqueryconnection/connection_mapping.go:298]: We'll have to figure out to do here.  We'll probably just have to skip the field.  That's acceptable I think.

File: pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/cloudsqlconnectionbasic/_http.log
Comment (justinsb) [inline pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/cloudsqlconnectionbasic/_http.log:781]: Uhoh ... should the password be provided in plaintext or base64?

Edit: nevermind, I see why, but I suggest using a password that doesn't cause confusion (like "password" or "secret")

File: pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/cloudsqlconnectionbasic/dependencies.yaml
Comment (justinsb) [inline pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/cloudsqlconnectionbasic/dependencies.yaml:21]: Nit: maybe make this `password: password` because otherwise people (OK, me!) will think you are forgetting to base64 the secret :-)

**Review by justinsb — APPROVED** [review]

**Comment by justinsb** [comment]
BTW I added the hold but looks like you've looked at the issues, so I think this is good to go @yuwenma

**Comment by yuwenma** [comment]
/hold cancel

Thank you for the reply @justinsb I'll resolve issues in the follow up PRs

**Pull request merged**
